pxl_scripts/px/redis_flow_graph/redis_flow_graph.pxl

```diff
@@ -15,52 +16,131 @@ window_ns = px.DurationNanos(10 * ns_per_s)
 
 def redis_flow_graph(start: str, ns: px.Namespace, from_entity_filter: str, to_entity_filter: str):
 
+    # redis_events.beta is a beta table, which is subject to change.
     df = px.DataFrame('redis_events.beta', start_time=start)
-    df.pod = df.ctx['pod']
-    df.namespace = df.ctx['namespace']
-    df = df.drop(['upid'])
+    df = add_source_dest_columns(df)
 
     # Filter on namespace as specified by the user.
     df = df[df.namespace == ns]
 
+    # Filter out messages with empty source / destination.
+    df = df[df.source != '']
+    df = df[df.destination != '']
+
+    # Filter out entities as specified by the user.
+    df = df[px.contains(df.source, from_entity_filter)]
+    df = df[px.contains(df.destination, to_entity_filter)]
+
     # Create 10 ns bin for time_ column
     df.timestamp = px.bin(df.time_, window_ns)
 
-    # Aggregate the connections by timestamp.
-    df = df.groupby(['timestamp', 'remote_addr', 'pod']).agg(
+    df = df.groupby(['timestamp', 'source', 'destination', 'is_source_pod_type',
+                     'is_dest_pod_type', 'namespace']).agg(
         latency_quantiles=('latency_ns', px.quantiles),
         throughput_total=('latency_ns', px.count),
     )
+
     df.latency_p50 = px.DurationNanos(px.floor(px.pluck_float64(df.latency_quantiles, 'p50')))
     df.latency_p90 = px.DurationNanos(px.floor(px.pluck_float64(df.latency_quantiles, 'p90')))
     df.latency_p99 = px.DurationNanos(px.floor(px.pluck_float64(df.latency_quantiles, 'p99')))
-
-    # Specify to and from entities.
-    df = df[df.remote_addr != '']
-    df.to_entity = df.pod
-    df.from_entity = px.pod_id_to_pod_name(px.ip_to_pod_id(df.remote_addr))
-
     df.request_throughput = df.throughput_total / window_ns
-    df = df.groupby(['from_entity', 'to_entity']).agg(
+
+    df = df.groupby(['source', 'destination', 'is_source_pod_type', 'is_dest_pod_type', 'namespace']).agg(
```

Comment:
I have an idea for how you can get around the "Mean of percentiles" problem but since it affects other diffs besides this one it's probably best done as a separate change

Reply:
Since this problem exists in many scripts, it'll be in a different diff. Is your idea something different than the UDF in PP-2491?

Reply:
yeah. We can add a UDF that does conversion from "-5m" to a time. (we've talked about it before). This removes the need to do 2 aggregates in a row. The reason we do 2 aggregates is basically trying to work around the fact that we don't have a way of converting the total time range to a time. (easier to explain over zoom if you have further questions)

another solution that doesn't require any UDFs is splitting the throughput calculation and the latency calculation into 2 separate tables. throughput will be the result of 2 different aggregates. Latency will be the result of 1 aggregate (since we don't need 2 aggregates for it like we do with throughput right now). Then, we join them together. This solution could work today without any compiler changes but it is a little confusing to the reader.